### PR TITLE
Empty trash bin

### PR DESCRIPTION
**Pull request by Aeredren**
fixes: https://github.com/nextcloud/tasks/issues/1692

### CODE REVIEW

File: src/components/AppNavigation/Trashbin.vue [hunk 1]
```diff
@@ -42,6 +42,14 @@ License along with this library. If not, see <http://www.gnu.org/licenses/>.
 					</EmptyContent>
 					<template v-else>
 						<h2>{{ t('tasks', 'Trash bin') }}</h2>
+						<ActionButton
```
Comment (raimund-schluessler) [inline src/components/AppNavigation/Trashbin.vue:45]: `ActionButton` needs to be wrapped inside and `Actions` component. See https://github.com/nextcloud/tasks/pull/1802/files#diff-a605b8ea178482e0a0bcab69e53d101699e820dde4ae75e617a1d993bd92c634R84-R92 for an example.

File: src/components/AppNavigation/Trashbin.vue [hunk 2]
```diff
@@ -42,6 +42,14 @@ License along with this library. If not, see <http://www.gnu.org/licenses/>.
...
+							@click="onEmptyTrashBin(items)">
+							<template #icon>
+								<Delete :size="20" decorative />
+							</template>
+							{{ t('tasks','Delete permanently') }}
```
Comment (raimund-schluessler) [inline src/components/AppNavigation/Trashbin.vue:50]:
```suggestion
							{{ t('tasks','Empty trash bin') }}
```

**Comment by Aeredren** [comment]
I don't really like what I did with the button. It should have a trashbin icon imo but I could'nt find how to add it. And perhaps it's better to put it at the end :
- less misclick possible
- but necessity to scroll down the loooong trashbin history to find it

Even better would be an undo action because we're talking batch deletion here... I will look toward it this week.

**Comment by raimund-schluessler** [comment]
@Aeredren Please don't commit changes to the l10n files. They are handled via transifex and generated automatically. Your changes would be overwritten on the next sync.

**Comment by raimund-schluessler** [comment]
> 
> 
> I don't really like what I did with the button. It should have a trashbin icon imo but I could'nt find how to add it.
 
Yeah, that's a bit tricky at the moment, you have to do it manually. Just put the `Delete` icon component within `button` and style the icon position properly. This will be better once we have the button component from nextcloud/vue: https://github.com/nextcloud/nextcloud-vue/pull/1808


> And perhaps it's better to put it at the end :
> 
>    * less misclick possible
> 
>    * but necessity to scroll down the loooong trashbin history to find it
> 

I don't have a strong opinion here. But if you put it at the end, it should be statically visible in the footer. Otherwise users will never find it.

> Even better would be an undo action because we're talking batch deletion here... I will look toward it this week.

Well, this is the trash bin. Having an undo for the trash bin would be weird, I think, since the trash bin is the undo already, no? We could have a confirmation modal though. That's also the way every other trash bin implementation I know of does it. And I know that undo is preferred over confirmation, but here we have one of the few exceptions, I would say. What do @nextcloud/designers think here?

**Comment by nimisha-vijay** [comment]
Are there any screenshots we could see?

But from what I can tell, I agree with everything, an "empty trash bin" button with a confirmation modal is good. Since it might get filled up with a lot of entries it makes sense to have it always visible. If there are 100+ entries, that's when you'll likely want to empty the bin anyway and you wouldn't want to scroll so much. A button that's fixed to the bottom of the modal sounds good.

**Comment by Aeredren** [comment]
I move the button to the footer, but I could not find how to add icons nor confirmation modal.
Here a screenshot of what it look like :
trashbin_screenshot1

For the icon we can maybe wait for the Button component ?
For the modal, if anyone know how to do it feel free to implement it.

**Comment by raimund-schluessler** [comment]
> For the modal, if anyone know how to do it feel free to implement it.

For the modal we will just use the Modal component from nextcloud/vue:
https://nextcloud-vue-components.netlify.app/#/Components/Modal

**Comment by raimund-schluessler** [comment]
> 
> 
> > For the modal, if anyone know how to do it feel free to implement it.
> 
> For the modal we will just use the Modal component from nextcloud/vue: https://nextcloud-vue-components.netlify.app/#/Components/Modal

Although the trash bin is a modal itself already, let's see how that works out 🙈

**Comment by Aeredren** [comment]
> Although the trash bin is a modal itself already, let's see how that works out see_no_evil

That's why I did'nt want to try and implement it, look like problems are coming ://
I might give it a try this afternoon if I have some time.

**Comment by raimund-schluessler** [comment]
> 
> 
> > Although the trash bin is a modal itself already, let's see how that works out see_no_evil
> 
> That's why I did'nt want to try and implement it, look like problems are coming :// I might give it a try this afternoon if I have some time.

It seems to be fine. Checkout this code:
```vue
<template>
	<div>
		<button @click="showModal1">Show Modal1</button>
		<modal v-if="modal1" @close="closeModal1" size="small">
			<div class="modal__content">Trash bin</div>
			<button @click="showModal2">Empty</button>
		</modal>
		<modal v-if="modal2" @close="closeModal2" size="small">
			<div class="modal__content">Your really want to do this?</div>
			<button>Yes</button>
			<button>No</button>
		</modal>
	</div>
</template>
<style scoped>
	.modal__content {
		margin: 50px;
		text-align: center;
	}

</style>
<script>
export default {
	data() {
		return {
			modal1: false,
			modal2: false,
		}
	},
	methods: {
		showModal1() {
			this.modal1 = true
		},
		closeModal1() {
			this.modal1 = false
		},
		showModal2() {
			this.modal2 = true
		},
		closeModal2() {
			this.modal2 = false
		}
	}
}
</script>
```
Copy it into the code input field on this page https://nextcloud-vue-components.netlify.app/#/Components/Modal to try it out.

**Comment by Aeredren** [comment]
I was feeling like reinventing the wheel with this confirmation modal. And as [no vue composant for confirmation exist yet](https://github.com/nextcloud/nextcloud-vue/issues/408) in nextcloud composant, I try to reuse the legacy OC.dialogs.confirm. Here what it look like :
trashbin_screenshot2

Is it a good idea to use this legacy code or not ? If not we could wait for the vue component or else I have a draft for the modal in enh/emptyTrash_modal. (But I'm not good at css ^^)

File: src/components/AppNavigation/Trashbin.vue [hunk 3]
```diff
@@ -251,6 +267,15 @@ export default {
 				showError(t('tasks', 'Could not restore calendar or event'))
 			}
 		},
+		async onEmptyTrashBin(items) {
```
Comment (raimund-schluessler) [inline src/components/AppNavigation/Trashbin.vue:270]:
```suggestion
		onEmptyTrashBin() {
			OC.dialogs.confirm(
				t('tasks', 'Do you really want to empty the trash bin?'),
				t('tasks', 'Empty trash bin'),
				this.emptyTrashBin,
				true
			)
		},
		emptyTrashBin(confirm) {
			if (!confirm) {
				return
			}
			this.items.forEach((item) => {
				this.onDeletePermanently(item)
			})
		},
```

File: src/components/AppNavigation/Trashbin.vue [hunk 4]
```diff
@@ -85,9 +85,14 @@ License along with this library. If not, see <http://www.gnu.org/licenses/>.
...
 							</template>
 						</div>
-						<p v-if="retentionDuration" class="footer">
-							{{ n('calendar', 'Elements in the trash bin are deleted after {numDays} day', 'Elements in the trash bin are deleted after {numDays} days', retentionDuration, { numDays: retentionDuration }) }}
-						</p>
+						<div class="footer">
+							<p v-if="retentionDuration">
+								{{ n('calendar', 'Elements in the trash bin are deleted after {numDays} day', 'Elements in the trash bin are deleted after {numDays} days', retentionDuration, { numDays: retentionDuration }) }}
+							</p>
+							<button @click="onEmptyTrashBin(items)">
```
Comment (raimund-schluessler) [inline src/components/AppNavigation/Trashbin.vue:92]:
```suggestion
							<button @click="onEmptyTrashBin()">
```

**Comment by raimund-schluessler** [comment]
> Is it a good idea to use this legacy code or not ? If not we could wait for the vue component or else I have a draft for the modal in [enh/emptyTrash_modal](https://github.com/Aeredren/tasks/tree/enh/emptyTrash_modal).

I don't really know. Other apps also still use OC.dialogs, so I guess it is fine.

I suggest to simplify the code a bit and create a second method instead of making the dialog a promise. I couldn't push to your branch, so please checkout the code suggestions above.

Besides that, it works nicely.

File: src/components/AppNavigation/Trashbin.vue (hunk 3 above)
Comment (Aeredren) [reply src/components/AppNavigation/Trashbin.vue]: Done.

File: src/components/AppNavigation/Trashbin.vue (hunk 4 above)
Comment (Aeredren) [reply src/components/AppNavigation/Trashbin.vue]: Done

**Comment by Aeredren** [comment]
I guess I'll port this to the Calendar trash bin once this PR is good enough to be merge.

**Comment by raimund-schluessler** [comment]
Works well now and the code looks good. 👍 

We should probably only run 5 delete requests at a time to not overload the server when many elements are to be deleted. But this can be done in a follow-up PR.

**Pull request merged**